R/jhu_data.R

```diff
@@ -283,3 +251,76 @@ load_jhu_data <- function(
 
   return(results)
 }
+
+
+#' Create a tibble of incident and cumulative deaths or cases based on a
+#' specified issue date and measure. This function will also download a
+#' specified time series data set if issue_date is not available in pre-built
+#' data object.
+#' 
+#' @param issue_date character issue date (i.e. report date) to use for
+#' constructing truths in format 'yyyy-mm-dd'
+#' @param as_of character issue date (i.e. report date) to use for
+#' constructing truths published as of this date in format 'yyyy-mm-dd'
+#' @param measure character vector specifying measure of covid prevalence:
+#' 'deaths' or 'cases'
+#' 
+#' @return tibble with issue_date and data
+#' 
+preprocess_jhu_data <- function(issue_date = NULL, as_of = NULL, measure = "deaths"){
+  
+  if (measure == "deaths"){
+    links <- covidData::jhu_deaths_data_links
+    jhu_data <- covidData::jhu_deaths_data
+    base_file_name <- "time_series_covid19_deaths_US.csv"
+  } else if (measure == "cases"){
+    links <- covidData::jhu_cases_data_links
+    jhu_data <- covidData::jhu_cases_data
+    base_file_name <- "time_series_covid19_confirmed_US.csv"
+  }
+  
+  # validate issue_date and as_of
+  if (!missing(issue_date) && !missing(as_of) &&
+      !is.null(issue_date) && !is.null(as_of)) {
+    stop("Cannot provide both arguments issue_date and as_of to load_jhu_data.")
+  } else if (is.null(issue_date) && is.null(as_of)) {
+    issue_date <- Sys.Date()
+  } else if (!is.null(as_of)) {
+    # case0: as_of < min(links$date) --> error
+    # case1: as_of >= min(links$date) 
+    #      a. as_of <= max --> get largest_issue_date <= as_of
+    #      b. as_of > max --> get the latest links, get largest_issue_date <= as_of
+    if (as_of < min(links$date)){
+      stop("Provided as_of date is earlier than all available issue dates.")
+    } else {
+      if (as_of > max(links$date)){
+        # query Github API to get the first page of results
+        links <- get_time_series_data_link(measure, first_page_only = TRUE)
+      }
+      issue_date <- max(links$date[links$date <= as.character(as_of)])
+    }
+  } else {
+    issue_date <- lubridate::ymd(issue_date)
+  }
+  
+  # subset jhu_data based on issue_date
+  if (issue_date %in% jhu_data$issue_date){
+    jhu_data <- jhu_data %>%
+      dplyr::filter(issue_date == UQ(issue_date))
+  } else {
+    if(!issue_date %in% links$date){
+      # query Github API to get the first page of results
+      links <- get_time_series_data_link(measure, first_page_only = TRUE)
```

Comment:
why do we specify `first_page_only = TRUE` here?  Would this work in a situation where:

- someone installed the package a long time ago, e.g. Jan 1 2021, and they have not re-installed since then
- They then ask for data with an issue date that's after Jan 1, but still pretty old, like Jan 2 2021.  This might not show up in the first page of updates from JHU right?

I'm wondering if there could be an option to pass the issue date we're after to the `get_time_series_data_link` function?